From b939c626902718aef03fddf57c675b27b81e3f3c Mon Sep 17 00:00:00 2001
From: Dean Sanner <dsanner@us.ibm.com>
Date: Tue, 19 Sep 2017 21:38:23 -0500
Subject: [PATCH v1] Correct NV link PLL calc, right justify lane mask

  - PLL buckets are counted one based, adjust to 0 based array
  - NV lane mask was in IBM bit 0..24, need to be 12..31 (24..0)

Change-Id: Ia2404b962acb9f2173c69faccb9cd48763bec0d9
---
 src/usr/hdat/hdatutil.C | 53 +++++++++++++++++++++++++------------------------
 1 file changed, 27 insertions(+), 26 deletions(-)

diff --git a/src/usr/hdat/hdatutil.C b/src/usr/hdat/hdatutil.C
index 1f3a322..8fdf9b8 100644
--- a/src/usr/hdat/hdatutil.C
+++ b/src/usr/hdat/hdatutil.C
@@ -43,38 +43,38 @@ extern trace_desc_t *g_trac_hdat;
 
 // SEQUOIA
 const hdatSMPLinkInfo_t l_hdatSMPLinkInfoProc0_6gpucfg[] = {
- {0,0x01,0x00,0xF1E00000,11,0xFFFF,0xFF,0x00,0xFF,0xFF,0xFF,2,22,0},
- {2,0x01,0x01,0x07187000,11,0xFFFF,0xFF,0x00,0xFF,0xFF,0xFF,2,22,0},
- {4,0x01,0x02,0x00078F00,13,0xFFFF,0xFF,0x00,0xFF,0xFF,0xFF,2,24,0},
- {6,0x01,0x09,0xF1E00000,15,0xFFFF,0xFF,0x00,0xFF,0xFF,0xFF,2,23,0},
- {8,0x01,0x0A,0x07187000,15,0xFFFF,0xFF,0x00,0xFF,0xFF,0xFF,2,23,0},
- {10,0x01,0x0B,0x00078F00,13,0xFFFF,0xFF,0x00,0xFF,0xFF,0xFF,2,24,0}
+ {0, 0x01,0x00,0xF1E000,11,0xFFFF,0xFF,0x00,0xFF,0xFF,0xFF,2,22,0},
+ {2, 0x01,0x01,0x0E1870,11,0xFFFF,0xFF,0x00,0xFF,0xFF,0xFF,2,22,0},
+ {4, 0x01,0x02,0x00078F,13,0xFFFF,0xFF,0x00,0xFF,0xFF,0xFF,2,24,0},
+ {6, 0x01,0x09,0xF1E000,13,0xFFFF,0xFF,0x00,0xFF,0xFF,0xFF,2,23,0},
+ {8, 0x01,0x0A,0x0E1870,15,0xFFFF,0xFF,0x00,0xFF,0xFF,0xFF,2,23,0},
+ {10,0x01,0x0B,0x00078F,15,0xFFFF,0xFF,0x00,0xFF,0xFF,0xFF,2,24,0},
 };
 const hdatSMPLinkInfo_t l_hdatSMPLinkInfoProc1_6gpucfg[] = {
- {0,0x01,0x00,0xF1E00000,24,0xFFFF,0xFF,0x00,0xFF,0xFF,0xFF,2,22,0},
- {2,0x01,0x01,0x07187000,24,0xFFFF,0xFF,0x00,0xFF,0xFF,0xFF,2,22,0},
- {4,0x01,0x02,0x00078F00,26,0xFFFF,0xFF,0x00,0xFF,0xFF,0xFF,2,23,0},
- {6,0x01,0x09,0xF1E00000,28,0xFFFF,0xFF,0x00,0xFF,0xFF,0xFF,2,24,0},
- {8,0x01,0x0A,0x07187000,28,0xFFFF,0xFF,0x00,0xFF,0xFF,0xFF,2,24,0},
- {10,0x01,0x0B,0x00078F00,26,0xFFFF,0xFF,0x00,0xFF,0xFF,0xFF,2,23,0}
+ {0, 0x01,0x00,0xF1E000,24,0xFFFF,0xFF,0x00,0xFF,0xFF,0xFF,2,22,0},
+ {2, 0x01,0x01,0x0E1870,24,0xFFFF,0xFF,0x00,0xFF,0xFF,0xFF,2,22,0},
+ {4, 0x01,0x02,0x00078F,26,0xFFFF,0xFF,0x00,0xFF,0xFF,0xFF,2,23,0},
+ {6, 0x01,0x09,0xF1E000,26,0xFFFF,0xFF,0x00,0xFF,0xFF,0xFF,2,24,0},
+ {8, 0x01,0x0A,0x0E1870,28,0xFFFF,0xFF,0x00,0xFF,0xFF,0xFF,2,24,0},
+ {10,0x01,0x0B,0x00078F,28,0xFFFF,0xFF,0x00,0xFF,0xFF,0xFF,2,23,0},
 };
 
 // REDBUD
 const hdatSMPLinkInfo_t l_hdatSMPLinkInfoProc0_4gpucfg[] = {
- {1,0x01,0x00,0xF1E00000,11,0xFFFF,0xFF,0x00,0xFF,0xFF,0xFF,2,22,0},
- {3,0x01,0x01,0x07187000,11,0xFFFF,0xFF,0x00,0xFF,0xFF,0xFF,2,22,0},
- {5,0x01,0x02,0x00078F00,11,0xFFFF,0xFF,0x00,0xFF,0xFF,0xFF,2,22,0},
- {7,0x01,0x09,0xF1E00000,13,0xFFFF,0xFF,0x00,0xFF,0xFF,0xFF,2,24,0},
- {9,0x01,0x0A,0x07187000,13,0xFFFF,0xFF,0x00,0xFF,0xFF,0xFF,2,24,0},
- {11,0x01,0x0B,0x00078F00,13,0xFFFF,0xFF,0x00,0xFF,0xFF,0xFF,2,24,0}
+ {1, 0x01,0x00,0xF1E000,11,0xFFFF,0xFF,0x00,0xFF,0xFF,0xFF,2,22,0},
+ {3, 0x01,0x01,0x0E1870,11,0xFFFF,0xFF,0x00,0xFF,0xFF,0xFF,2,22,0},
+ {5, 0x01,0x02,0x00078F,11,0xFFFF,0xFF,0x00,0xFF,0xFF,0xFF,2,22,0},
+ {7, 0x01,0x09,0xF1E000,13,0xFFFF,0xFF,0x00,0xFF,0xFF,0xFF,2,24,0},
+ {9, 0x01,0x0A,0x0E1870,13,0xFFFF,0xFF,0x00,0xFF,0xFF,0xFF,2,24,0},
+ {11,0x01,0x0B,0x00078F,13,0xFFFF,0xFF,0x00,0xFF,0xFF,0xFF,2,24,0}
 };
 const hdatSMPLinkInfo_t l_hdatSMPLinkInfoProc1_4gpucfg[] = {
- {1,0x01,0x00,0xF1E00000,24,0xFFFF,0xFF,0x00,0xFF,0xFF,0xFF,2,22,0},
- {3,0x01,0x01,0x07187000,24,0xFFFF,0xFF,0x00,0xFF,0xFF,0xFF,2,22,0},
- {5,0x01,0x02,0x00078F00,24,0xFFFF,0xFF,0x00,0xFF,0xFF,0xFF,2,22,0},
- {7,0x01,0x09,0xF1E00000,26,0xFFFF,0xFF,0x00,0xFF,0xFF,0xFF,2,23,0},
- {9,0x01,0x0A,0x07187000,26,0xFFFF,0xFF,0x00,0xFF,0xFF,0xFF,2,23,0},
- {11,0x01,0x0B,0x00078F00,26,0xFFFF,0xFF,0x00,0xFF,0xFF,0xFF,2,23,0}
+ {1, 0x01,0x00,0xF1E000,24,0xFFFF,0xFF,0x00,0xFF,0xFF,0xFF,2,22,0},
+ {3, 0x01,0x01,0x0E1870,24,0xFFFF,0xFF,0x00,0xFF,0xFF,0xFF,2,22,0},
+ {5, 0x01,0x02,0x00078F,24,0xFFFF,0xFF,0x00,0xFF,0xFF,0xFF,2,22,0},
+ {7, 0x01,0x09,0xF1E000,26,0xFFFF,0xFF,0x00,0xFF,0xFF,0xFF,2,23,0},
+ {9, 0x01,0x0A,0x0E1870,26,0xFFFF,0xFF,0x00,0xFF,0xFF,0xFF,2,23,0},
+ {11,0x01,0x0B,0x00078F,26,0xFFFF,0xFF,0x00,0xFF,0xFF,0xFF,2,23,0}
 };
 
 
@@ -2088,8 +2088,9 @@ errlHndl_t hdatUpdateSMPLinkInfoData(hdatHDIFDataArray_t * i_SMPInfoFullPcrdHdrP
                             0,0,0,0);
                     break;
             }
-            
-            uint32_t l_pllfreq = *(l_freqList+l_obusPllFreqBucket);
+
+            //PLL bucket is 1 based (1,2,3), subtract 1 for 0 based array
+            uint32_t l_pllfreq = *(l_freqList+l_obusPllFreqBucket -1);
             
             switch( l_pllfreq ){
                 case 1250:{l_SMPInfoEle.hdatSMPLinkSpeed = HDAT_OBUS_FREQ_20GBPS; break; };
-- 
1.8.2.2

